ctapipe/reco/shower_processor.py

```diff
@@ -0,0 +1,193 @@
+        classify,
+        config=None,
+        parent=None,
+        **kwargs,
+    ):
+        """
+        Parameters
+        ----------
+        subarray: SubarrayDescription
+            Description of the subarray. Provides information about the
+            camera which are useful in calibration. Also required for
+            configuring the TelescopeParameter traitlets.
+        is_simulation: bool
+            If true, also process simulated images if they exist
+        config: traitlets.loader.Config
+            Configuration specified by config file or cmdline arguments.
+            Used to set traitlet values.
+            This is mutually exclusive with passing a ``parent``.
+        parent: ctapipe.core.Component or ctapipe.core.Tool
+            Parent of this component in the configuration hierarchy,
+            this is mutually exclusive with passing ``config``
+        """
+
+        super().__init__(config=config, parent=parent, **kwargs)
+        self.subarray = subarray
+        self.check_shower = ShowerQualityQuery(parent=self)
+        self._is_simulation = is_simulation
+        self.reconstruct_energy = reconstruct_energy
+        self.classify = classify
+        self.reconstructor = HillasReconstructor()
+
+    def _reconstruct_shower(
+        self,
+        event,
+        default=DEFAULT_SHOWER_PARAMETERS,
+    ) -> ReconstructedShowerContainer:
+        """Perform shower reconstruction.
+
+        Parameters
+        ----------
+        tel_id: int
+            which telescope is being cleaned
+        image: np.ndarray
+            image to process
+        signal_pixels: np.ndarray[bool]
+            image mask
+        peak_time: np.ndarray
+            peak time image
+        Returns
+        -------
+        ReconstructedShowerContainer:
+            direction in the sky with uncertainty,
+            core position on the ground with uncertainty,
+            h_max with uncertainty,
+            is_valid boolean for successfull reconstruction,
+            average intensity of the intensities used for reconstruction,
+            measure of algorithm success (if fit),
+            list of tel_ids used if stereo, or None if Mono
+        """
+
+        # Read only valid HillasContainers (minimum condition to continue)
+        hillas_dict = {
+            tel_id: dl1.parameters.hillas
+            for tel_id, dl1 in event.dl1.tel.items()
+            if np.isfinite(event.dl1.tel[tel_id].parameters.hillas.intensity)
+        }
+
+        # On top of this check if the shower should be considered based
+        # on the user's configuration
+        shower_criteria = [self.check_shower(hillas_dict[tel_id]) for tel_id in hillas_dict]
+        self.log.debug(
+            "shower_criteria: %s",
+            list(zip(self.check_shower.criteria_names[1:], shower_criteria)),
+        )
+
+        # Reconstruct the shower only if all shower criteria are met
+        if np.count_nonzero(shower_criteria) > 2:
```

Comment:
You check for at least two telescopes surviving the criteria, but you use all available telescopes here!

I find that surprising. I would expect only those telescopes are used for stereo reconstruction where the requirements are met.

Reply:
ah right.

maybe a mask on hillas_dict?

Reply:
No, just use the quality criterion to create the hillas dict

Reply:
ah this is what you meant with the comment above...

Reply:
might be clearer to first build a list of "good" tel_ids insted of a list of shower criteria (which is not indexed by id currently). Something on the lines of the following (though there are multple ways to do it):
```py3
quality_by_tel = {tel_id: self.check_shower(hillas_dict[tel_id]) for tel_id in hillas_dict}
good_tel_ids = [tid for tid,quality in quality_by_tel.items() if all(quality)]
```

Another comment is that you should not apply `check_shower` to  hillas_dict, but to `event.dl1.tel[i].parameters`. That allows one to make selections on *any* parameter, not just hillas parameters.

Reply:
> Another comment is that you should not apply check_shower to hillas_dict, but to event.dl1.tel[i].parameters. That allows one to make selections on any parameter, not just hillas parameters.

yes in fact to use n_pixels I need to do this

Reply:
done